mongo/client_bulk_write.go

```diff
@@ -0,0 +1,706 @@
+	fn := functionSet{
+		appendStart:    bsoncore.AppendArrayElementStart,
+		appendDocument: bsoncore.AppendDocumentElement,
+		updateLength: func(dst []byte, idx, _ int32) []byte {
+			dst, _ = bsoncore.AppendArrayEnd(dst, idx)
+			return dst
+		},
+	}
+	return mb.appendBatches(fn, dst, maxCount, maxDocSize, totalSize)
+}
+
+type functionSet struct {
+	appendStart    func([]byte, string) (int32, []byte)
+	appendDocument func([]byte, string, []byte) []byte
+	updateLength   func([]byte, int32, int32) []byte
+}
+
+func (mb *modelBatches) appendBatches(fn functionSet, dst []byte, maxCount, maxDocSize, totalSize int) (int, []byte, error) {
+	if mb.Size() == 0 {
+		return 0, dst, io.EOF
+	}
+
+	mb.cursorHandlers = mb.cursorHandlers[:0]
+	mb.newIDMap = make(map[int]interface{})
+
+	nsMap := make(map[string]int)
+	getNsIndex := func(namespace string) (int, bool) {
+		v, ok := nsMap[namespace]
+		if ok {
+			return v, ok
+		}
+		nsIdx := len(nsMap)
+		nsMap[namespace] = nsIdx
+		return nsIdx, ok
+	}
+
+	canRetry := true
+	checkSize := true
+
+	l := len(dst)
+
+	opsIdx, dst := fn.appendStart(dst, "ops")
+	nsIdx, nsDst := fn.appendStart(nil, "nsInfo")
+
+	totalSize -= 1000
+	size := len(dst) + len(nsDst)
+	var n int
+	for i := mb.offset; i < len(mb.models); i++ {
+		if n == maxCount {
+			break
+		}
+
+		ns := mb.models[i].namespace
+		nsIdx, exists := getNsIndex(ns)
+
+		var doc bsoncore.Document
+		var err error
+		switch model := mb.models[i].model.(type) {
+		case *ClientInsertOneModel:
+			checkSize = false
+			mb.cursorHandlers = append(mb.cursorHandlers, mb.appendInsertResult)
+			var id interface{}
+			id, doc, err = (&clientInsertDoc{
+				namespace: nsIdx,
+				document:  model.Document,
+				sizeLimit: maxDocSize,
+			}).marshal(mb.client.bsonOpts, mb.client.registry)
+			if err != nil {
+				break
+			}
+			mb.newIDMap[i] = id
+		case *ClientUpdateOneModel:
+			mb.cursorHandlers = append(mb.cursorHandlers, mb.appendUpdateResult)
+			doc, err = (&clientUpdateDoc{
+				namespace:      nsIdx,
+				filter:         model.Filter,
+				update:         model.Update,
+				hint:           model.Hint,
+				arrayFilters:   model.ArrayFilters,
+				collation:      model.Collation,
+				upsert:         model.Upsert,
+				multi:          false,
+				checkDollarKey: true,
+			}).marshal(mb.client.bsonOpts, mb.client.registry)
+		case *ClientUpdateManyModel:
+			canRetry = false
+			mb.cursorHandlers = append(mb.cursorHandlers, mb.appendUpdateResult)
+			doc, err = (&clientUpdateDoc{
+				namespace:      nsIdx,
+				filter:         model.Filter,
+				update:         model.Update,
+				hint:           model.Hint,
+				arrayFilters:   model.ArrayFilters,
+				collation:      model.Collation,
+				upsert:         model.Upsert,
+				multi:          true,
+				checkDollarKey: true,
+			}).marshal(mb.client.bsonOpts, mb.client.registry)
+		case *ClientReplaceOneModel:
```

Comment:
From the specifications:

> Replacement documents provided in `ReplaceOne` write models are required not to contain atomic modifiers. Drivers MUST throw an error if a replacement document is nonempty and it's first key starts with `$`.

The following does not error as expected:

```go
package main

import (
	"context"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
)

func main() {
	client, err := mongo.Connect()
	if err != nil {
		panic(err)
	}

	defer func() { _ = client.Disconnect(context.Background()) }()

	pairs := &mongo.ClientWriteModels{}

	insertOneModel := mongo.NewClientInsertOneModel().SetDocument(bson.D{{"x", 1}})

	pairs = pairs.AppendInsertOne("db", "k", insertOneModel)
	_, err = client.BulkWrite(context.Background(), pairs) // Should not panic
	if err != nil {
		panic(err)
	}

	pairs = &mongo.ClientWriteModels{}

	replaceOneModel := mongo.NewClientReplaceOneModel().SetFilter(bson.D{}).SetReplacement(bson.D{{"$x", 1}})

	pairs = pairs.AppendReplaceOne("db", "k", replaceOneModel)

	_, err = client.BulkWrite(context.Background(), pairs) // Should not panic
	if err != nil {
		panic(err)
	}
}
```

Suggest adding a unit test and updating the unified spec tests to include this case.

Reply:
We can rely on the server to throw the error.

Reply:
Disregard this comment, this functionality works as expected.